### PR TITLE
Implement methods to adding/removing chains to a LocalRegistry

### PR DESCRIPTION
**Pull request by jmrossy**

### CODE REVIEW

File: src/registry/BaseRegistry.ts [hunk 1]
```diff
@@ -36,4 +36,11 @@ export abstract class BaseRegistry implements IRegistry {
   abstract getChainMetadata(chainName: ChainName): MaybePromise<ChainMetadata | null>;
   abstract getAddresses(): MaybePromise<ChainMap<ChainAddresses>>;
   abstract getChainAddresses(chainName: ChainName): MaybePromise<ChainAddresses | null>;
+  abstract addChains(
```
Comment (yorhodes) [inline src/registry/BaseRegistry.ts:39]: should we have single chain interface to mirror the `get`?

File: src/registry/LocalRegistry.ts [hunk 2]
```diff
@@ -89,6 +91,65 @@ export class LocalRegistry extends BaseRegistry implements IRegistry {
...
   }
 
+  addChains(
+    chains: Array<{ chainName: ChainName; metadata?: ChainMetadata; addresses?: ChainAddresses }>,
+  ): void {
+    const currentChains = this.listRegistryContent().chains;
+    for (const chain of chains) {
+      if (currentChains[chain.chainName])
+        throw new Error(`Chain ${chain.chainName} already exists in registry`);
+    }
```
Comment (yorhodes) [inline src/registry/LocalRegistry.ts:101]: should we throw or merge here?

File: src/registry/LocalRegistry.ts [hunk 3]
```diff
@@ -99,4 +160,37 @@ export class LocalRegistry extends BaseRegistry implements IRegistry {
...
+  protected createChainFile(
+    chainName: ChainName,
+    fileName: keyof ChainFiles,
+    data: any,
+    cache: ChainMap<any>,
+    prefix?: string,
+  ) {
+    const filePath = path.join(this.uri, this.getChainsPath(), chainName, `${fileName}.yaml`);
+    this.listContentCache!.chains[chainName] ||= {};
+    this.listContentCache!.chains[chainName][fileName] = filePath;
```
Comment (yorhodes) [inline src/registry/LocalRegistry.ts:182]: why is this definitely defined here?

File: src/registry/LocalRegistry.ts [hunk 4]
```diff
@@ -99,4 +160,37 @@ export class LocalRegistry extends BaseRegistry implements IRegistry {
...
+    cache[chainName] = data;
+    this.createFile({ filePath, data: toYamlString(data, prefix) });
+  }
+
+  protected removeFiles(filePaths: string[]): void {
+    for (const filePath of filePaths) {
+      fs.unlinkSync(filePath);
```
Comment (yorhodes) [inline src/registry/LocalRegistry.ts:189]: is this a lock?

File: src/registry/LocalRegistry.ts [hunk 5]
```diff
@@ -89,6 +91,65 @@ export class LocalRegistry extends BaseRegistry implements IRegistry {
...
+        this.createChainFile(
+          chainName,
+          'metadata',
+          chain.metadata,
+          this.metadataCache!,
+          CHAIN_SCHEMA_REF,
+        );
+      }
+      if (chain.addresses) {
+        this.createChainFile(chainName, 'addresses', chain.addresses, this.addressCache!);
```
Comment (yorhodes) [inline src/registry/LocalRegistry.ts:133]: would we ever add an addresses schema later?

File: src/registry/LocalRegistry.ts [hunk 6]
```diff
@@ -89,6 +91,65 @@ export class LocalRegistry extends BaseRegistry implements IRegistry {
...
+    this.getAddresses();
+
+    for (const chain of chains) {
+      const chainName = chain.chainName;
+      if (chain.metadata) {
+        this.createChainFile(
+          chainName,
+          'metadata',
+          chain.metadata,
+          this.metadataCache!,
```
Comment (yorhodes) [inline src/registry/LocalRegistry.ts:128]: slightly dont like this block, could we have the cache be returned in a var or have createChainFile read from the cache?

File: src/registry/GithubRegistry.ts [hunk 7]
```diff
@@ -123,6 +123,20 @@ export class GithubRegistry extends BaseRegistry implements IRegistry {
...
+    throw new Error('TODO: Implement');
+  }
+  async updateChains(
+    _chains: Array<{ chainName: ChainName; metadata?: ChainMetadata; addresses?: ChainAddresses }>,
+  ): Promise<void> {
+    throw new Error('TODO: Implement');
+  }
+  async removeChains(_chains: Array<ChainName>): Promise<void> {
+    throw new Error('TODO: Implement');
+  }
```
Comment (yorhodes) [inline src/registry/GithubRegistry.ts:138]: honestly not sure if we will ever implement this now that I think about it

Comment (yorhodes) [reply src/registry/GithubRegistry.ts:138]: we probably want to do the syncing all at once

**Review by yorhodes — APPROVED** [review]
lgtm, not sure if github registry should have the same `write` interface

File: src/registry/LocalRegistry.ts (hunk 4 above)
Comment (jmrossy) [reply src/registry/LocalRegistry.ts:189]: No, this is how you delete a file with node's fs lib. Strange I know.

File: src/registry/LocalRegistry.ts (hunk 2 above)
Comment (jmrossy) [reply src/registry/LocalRegistry.ts:101]: IMO people should use `updateChain` if they want add-or-merge behavior

File: src/registry/LocalRegistry.ts (hunk 5 above)
Comment (jmrossy) [reply src/registry/LocalRegistry.ts:133]: We can if you'd like, we just need a set list of contract names. Gonna punt on it for now

File: src/registry/LocalRegistry.ts (hunk 6 above)
Comment (jmrossy) [reply src/registry/LocalRegistry.ts:128]: Reworked it a bit, thanks for the suggestion

**Pull request merged**